### ISSUE TITLE
CoqIDE in v8.14 / master does not run on Windows

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#14800]
I built the v8.14 branch on native Windows using coq platform (tweaking 3 files in the platform).  When I run the resulting CoqIDE, I see the following. `Reset` immediately brings up the same dialog.  Debug output at bottom.  FWIW, note the CoqIDE version `8.14+alpha` in the Messages panel.

If I build v8.13.2 similarly, CoqIDE comes up just fine.

Perhaps someone else could try building and running this?  (At the moment, I can't run CoqIDE on Windows with `dune exec -- dev/shim/coqide-prelude` either.  More about that later.)

![image](https://user-images.githubusercontent.com/1253341/130371099-455474c8-4b42-4dd0-aa11-8d5a09c5d72f.png)

```
$ coqide -debug
(Spawn  ,616) EXEC: C:\cygwin64_coq2\home\jfehr\.opam\_coq-platform_.2021.03.0\bin\coqidetop.opt.exe -main-channel 127.0.0.1:52753:52754 -control-channel 127.0.0.1:52751:52752 --xml_format=Ppcmds -async-proofs off -debug
[DEBUG] Start eval_call Init None
[DEBUG] End eval_call
[DEBUG] End of Coqide.main
[DEBUG] Signal ignored (normal if Win32)
[DEBUG] Coqtop reader failed, resetting: Coq.TubeError("ERR")
Gtk-Message: GtkDialog mapped without a transient parent. This is discouraged.
[pid 204] <-- Init None
[pid 204] --> GOOD 1
```

**Comment by Zimmi48** [comment]
FWIW @MSoegtropIMC intends to extend the support of the Coq platform to earlier Coq versions. I don't know how much progress he has made on that front, but this could be of help.

---

Regarding triggering CI builds on old commits, no you don't need to create a GitHub App or anything like that. Which exact CI job are you trying to run?

Remember that there have been a number of CI providers used by the Coq repository over the years. In your `v814ancestor` branch, there is no GitHub CI configuration file (because at the time, we didn't use GitHub Actions), so nothing gets started, except for an unrelated `check-conflict` workflow. In case you merge with `master`, then a build gets started, because there is a GitHub CI configuration file...

If your goal is to obtain a Windows installer, then I expect that the CI job you want is actually run on GitLab. In this case, your best shot is to put the commit you want to build to a new branch on a GitLab repository. But there are chances that this will not work as is, because CI breaks often and needs fixes which may not be present in earlier versions.

If bisecting which commit created the issue takes that much effort, I wouldn't bother, if I were you...

**Comment by jfehrle** [comment]
I expect we want 8.14 to support Windows. Who else can work the issue?  In addition, the platform seems to be broken for Windows since Aug 22 because of recent Cygwin updates (if you're installing  Cygwin).  If I use an already-installed Cygwin, I can still build the 2021.02 platform, but if I try to build my variant modified to point to 8.13, it fails, either not creating CoqIDE or not able to start CoqIDE.

I'll be leaving town shortly.  Perhaps worth mentioning in next week's Coq call?

**Comment by jfehrle** [comment]
PS v814ancestor is from about a year ago.  Has the build infrastructure changed that much since then?

**Comment by Zimmi48** [comment]
>Has the build infrastructure changed that much since then?

Apparently yes.

>Who else can work the issue?

I expect @MSoegtropIMC will make sure everything is in order with the platform, to build 8.13, 8.14 and possibly even earlier versions.

**Comment by MSoegtropIMC** [comment]
I am busy making Coq Platform working with opam 2.1.0 (which is tricky cause of the depext changes).

Regarding the multi-version variant: it should just need the latest cygwin patch. You can find it here : (https://github.com/MSoegtropIMC/platform/tree/multi-version). I will push this in the next days (assuming I get the opam stuff fixed) as new master branch of Coq Platform and do a new release based on that (which has the advantage that one can just add 8.14 later).

**Comment by jfehrle** [comment]
Using the updated 2021.02 as a base and pointing to Coq `V8.14+alpha`, I don't get a `coqide` executable.

If I point to Coq `v8.14`, I get a runnable coqide but it still fails as I described at the top of this page.

What does the multi-version variant do?  This is the first time I've seen it mentioned.

**Comment by MSoegtropIMC** [comment]
The multi-variant supports multiple Coq variants and the plan is to support a 8.14 preview (8.14 is not yet released, so it can't support 8.14).

**Comment by jfehrle** [comment]
Bisection indicates that the failure began with
`[f590493defdae8a84e6c88ac981f89ac33744891] [build] Use dune to build OCaml code.`

That has 125 changed files, so there's still some work to do.

`make clean` was not useful.  It got errors due to Windows' command line length limit and probably isn't robust enough in principle for `git bisect`  I had to do this for each step:

```
rm -rf *
git checkout -f
./configure -profile devel
make -j 12 world
./_build/install/default/bin/coqide.exe -boot -noinit -debug
   OR
bin/coqide -boot -noinit
```

Just for the record:

```
$ git bisect bad
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[f590493defdae8a84e6c88ac981f89ac33744891] [build] Use dune to build OCaml code.
$ git bisect log
git bisect start
# bad: [91a89cb01d39c7329b6bc72648da1d3577f48f82] Backport PR #14855: Mark #14610 and #14664 as present in 8.14.
git bisect bad 91a89cb01d39c7329b6bc72648da1d3577f48f82
# good: [34da65afbe8fec419498f113548a7a58638cf404] update version to 8.13.2
git bisect good 34da65afbe8fec419498f113548a7a58638cf404
# good: [04b25a7635e796ad05ef7db537883594a5144a56] Merge PR #12873: Unification: A type-checking fix in imitation + an error locating fix
git bisect good 04b25a7635e796ad05ef7db537883594a5144a56
# good: [22be1892ee1b6030cfe9406bf72bb320dbe68be7] Slightly richer API allowing to shift the inductive in a block.
git bisect good 22be1892ee1b6030cfe9406bf72bb320dbe68be7
# bad: [a020c80788aa28a4b91bd45d81325c9c07b497a0] respect debug flag in Ltac2.Std.eauto
git bisect bad a020c80788aa28a4b91bd45d81325c9c07b497a0
# good: [d230b696a8b476e1c86f00a3ca5631ab6db83e40] Merge PR #14127: Pin docutils to 0.16.
git bisect good d230b696a8b476e1c86f00a3ca5631ab6db83e40
# bad: [2cd6214d59d041430ad7a2ffd8ddf09f7aacf6be] Merge PR #14206: [build] Fix a missing EXE suffix in makefiles
git bisect bad 2cd6214d59d041430ad7a2ffd8ddf09f7aacf6be
# good: [d9e9a63f9f49768eee8b239812365ad1115b964f] Merge PR #14075: New level of abstraction for streams with (non-canonical) location function
git bisect good d9e9a63f9f49768eee8b239812365ad1115b964f
# bad: [0727695ee4dcc293738d4c755d1f61c8f7ca4e5b] Merge PR #13617: [build] Use dune to build OCaml code.
git bisect bad 0727695ee4dcc293738d4c755d1f61c8f7ca4e5b
# good: [f7e05393addb9a45fa1eaef2a72da9b2c4c14396] Merge PR #13892: Move equality sections out of Tactics chapter.
git bisect good f7e05393addb9a45fa1eaef2a72da9b2c4c14396
# good: [f709cafaacf9a2c64a69f7a06a66d09e8a2b0309] Merge PR #14101: Cycle detection algorithm for universes : use a better criterion for determining the size of levels.
git bisect good f709cafaacf9a2c64a69f7a06a66d09e8a2b0309
# good: [e587915367b8e314ad73d329b12da4cd2ebe0de8] Merge PR #13751: Fix order of aliases when reasoning up to aliases in unification
git bisect good e587915367b8e314ad73d329b12da4cd2ebe0de8
# bad: [fcbdd1527a2a7e200d04257115ec37c9fe14eb3a] [ci] Temporal redirection of upstream gappa/interval
git bisect bad fcbdd1527a2a7e200d04257115ec37c9fe14eb3a
```

**Comment by jfehrle** [comment]
@ejgallego My best guess is that something was lost in the changes to `Makefile.ide`.

**Comment by MSoegtropIMC** [comment]
My guess is that it is a path issue caused by the cygwin MinGW cross compilation - similar to the issues that the shims don't work.

Which suggests that things might work if one provides an explicit correct coqlib path or uses the configuration file feature introduced by @gares - actually we should check what is in there ...

**Comment by gares** [comment]
This lines from ide/coqide/dune are suspicious
```
(rule
 (targets coqide_os_specific.ml)
 (deps (:in-file coqide_X11.ml.in)) ; TODO support others
 (action (copy# %{in-file} %{targets})))
```

**Comment by gares** [comment]
The windows specific file does thinker with the path, while the OSX one seems to better integrate the app with the top bar of the desktop.

**Comment by gares** [comment]
@ejgallego is it hard to select a file based on the OS in dune?
I know how to select a file based on the presence of a library, but not on the os.

**Comment by gares** [comment]
maybe we can use
```
(deps (:in-file coqide_%{os_type}.ml.in))
```
and rename files accordingly, but I did not try.

**Comment by ejgallego** [comment]
That's bizarre tho, the code is the same in master w.r.t. that file selection and it works.

OS-specific magic is happening in the OCaml code these days, using `Sys.os_type` I recall.

**Comment by ejgallego** [comment]
But sure, update that line and try, if that fixes the problem I will fix the dune file.

**Comment by Zimmi48** [comment]
> That's bizarre tho, the code is the same in master w.r.t. that file and it works.

Was it actually tested to work in `master`?

**Comment by gares** [comment]
> OS-specific magic is happening in the OCaml code these days, using `Sys.os_type` I recall.

Some stuff can't. For example in OSX we call code from a lablgtk module which does not exist on other platforms, so we can't really put it behind a runtime if.

**Comment by jfehrle** [comment]
> Which suggests that things might work if one provides an explicit correct coqlib path

If I specify `-boot -noinit` (which I did) then I think Coq doesn't look for coqlib until a proof is started.  I will try the suggestions given here shortly.

**Comment by ejgallego** [comment]
@jfehrle can you check the artifact in #12425 [once it is created] ?

**Comment by ejgallego** [comment]
That's the link to the artifact https://github.com/coq/coq/suites/3725561051/artifacts/91197841

Gonna update the PR.

**Comment by jfehrle** [comment]
Using the version of #12425 I see the same CoqIDE problem.

A few off-topic details:

1. Would be nice to get Windows to recognize Coq or Inria as a publisher: 

![image](https://user-images.githubusercontent.com/1253341/132728575-6f1c2558-7a05-445e-9785-e26c581e9f74.png)

2. On the console I see these messages.  There is no `gdk-pixbuf-2.0` directory in `C:\Coq8.15.alpha\lib`

```
(coqide.exe:42736): GdkPixbuf-WARNING **: Cannot open pixbuf loader module file 'C:\Coq8.15.alpha\lib\gdk-pixbuf-2.0\2.10.0\loaders.cache': No such file or directory

This likely means that your installation is broken.
Try running the command
  gdk-pixbuf-query-loaders > C:\Coq8.15.alpha\lib\gdk-pixbuf-2.0\2.10.0\loaders.cache
to make things work again for the time being.
```

**Comment by MSoegtropIMC** [comment]
@jfehrle : official releases are signed by INRIA, but test builds not (it is a manual process).

**Comment by ejgallego** [comment]
@jfehrle sorry, I messed the PR number, please try the link I provided, the correct PR number is in the corrected message.

**Comment by jfehrle** [comment]
I get the same behavior using https://github.com/coq/coq/suites/3725561051/artifacts/91197841

**Comment by jfehrle** [comment]
But I guess you want me to try #14873 once it's built the installer, right?

**Comment by ejgallego** [comment]
That link was for the #14873 build, tho it is strange that build succeeded, without having declared the stub, ummm.

**Comment by jfehrle** [comment]
Running `./_build/install/default/bin/coqide.exe -coqlib "$(cygpath -aw .)"\_build\default` I get this error dialog.  What should I have for `coqlib` instead?  (This command worked for several of the intermediate versions I built while bisecting before I switched to `-boot -noinit`.  I don't think this is the reason CoqIDE is broken.)

![image](https://user-images.githubusercontent.com/1253341/132752361-3733e891-b8c5-429d-850f-153a50d26107.png)

I see the .vos file here but `-coqlib "$(cygpath -aw .)"/_build_vo/default` doesn't work, either

```
$ find . -name Prelude.vos
./_build_vo/default/theories/Init/Prelude.vos
```

![image](https://user-images.githubusercontent.com/1253341/132753069-be0ed870-7f22-4b4f-b773-ee53c7a58775.png)

**Comment by jfehrle** [comment]
Great!  The latest build from #14873 works.  Let's get that merged ASAP.

Much less important: As I mentioned earlier, I still see this message in the Windows command window when I start coqide.  No idea if it makes a difference or is just noise.

```
c:\Coq8.15alpha\bin>coqide.exe

(coqide.exe:56904): GdkPixbuf-WARNING **: Cannot open pixbuf loader module file 'c:\Coq8.15alpha\lib\gdk-pixbuf-2.0\2.10.0\loaders.cache': No such file or directory

This likely means that your installation is broken.
Try running the command
  gdk-pixbuf-query-loaders > c:\Coq8.15alpha\lib\gdk-pixbuf-2.0\2.10.0\loaders.cache
to make things work again for the time being.
```

When I run the suggested command I get

```
c:\Coq8.15alpha\bin>gdk-pixbuf-query-loaders > c:\Coq8.15alpha\lib\gdk-pixbuf-2.0\2.10.0\loaders.cache
The system cannot find the path specified.
```

**Comment by MSoegtropIMC** [comment]
Regarding the loader cache: as long as you see icons I wouldn't worry too much about it. The `gdk-pixbuf-query-loaders` file can only be started from cygwin - not from Windows. It might be that the location of the cache file changed so that the installer misses it.

I guess the above report is from an installed Windows.

**Comment by jfehrle** [comment]
> I guess the above report is from an installed Windows.

From CoqIDE installed with the windows installer (and built in CI by #14873).

If we can't get the messages to go away, the release notes can mention that they don't matter.  (We _do_ have release notes somewhere, don't we?)

**Comment by Zimmi48** [comment]
People rarely run the CoqIDE executable from the command-line though.

**Comment by jfehrle** [comment]
In this case, I'm not running from the command line.  If I double-click on the executable in the File Manger, a command window pops up just before the CoqIDE window appears.  Same thing if I pin it to the task bar.

This is a change in behavior; my old 8.13.0 CoqIDE doesn't bring up the command window.  Nor does the installer prompt to add Coq to the task bar--I thought it used to do that.

FWIW, If I explicitly add 8.15alpha to the task bar, the set of info panels on their property dialogs is different.

![image](https://user-images.githubusercontent.com/1253341/132911189-de8fb34c-5d9c-4559-ae4f-1282fa6fd40d.png)

**Comment by jfehrle** [comment]
Also note that I don't see the pixbuf messages when I build locally and run from the command line.

**Comment by gares** [comment]
The icon is probably a missing link of ide/coqide/coq_icon.o (there is some code to call windres in makefile.ide)

**Comment by Zimmi48** [comment]
So is this another regression on the Coq side?

**Comment by jfehrle** [comment]
Asking me?  The Coq side as opposed to CoqIDE?  It is certainly in the CoqIDE process; Coq has no need for pixel buffers.

**Comment by Zimmi48** [comment]
Asking around. The Coq repository as opposed to the Platform infrastructure.

**Comment by gares** [comment]
i guess so.

**Comment by Zimmi48** [comment]
Maybe then it would deserve another issue so that it is not forgotten about when this bug is fixed.

**Comment by MSoegtropIMC** [comment]
Is this now fixed in the 8.14 release branch and master, or only in one of them?

**Comment by gares** [comment]
Not yet, see https://github.com/coq/coq/projects/45
